Methods and tests for jep334 String

theresa-m opened the pull request:
- Java12andUp test framework
- String.describeConstable & tests
- String.resolveConstantDesc & tests

Depends on (see https://github.com/eclipse/openj9/issues/4195): 
- Fix Java 12 build failures: https://github.com/eclipse/openj9/pull/3980
- Constants API stubs: https://github.com/eclipse/openj9/pull/4030
- Java12andUp test setup: https://github.com/eclipse/openj9/pull/4104

Signed-off-by: Theresa Mammarella <Theresa.T.Mammarella@ibm.com>

pshipton commented:
String implements Constable, ConstantDesc
http://cr.openjdk.java.net/~iris/se/12/build/latest/api/java.base/java/lang/String.html

theresa-m commented:
implements Constable, ConstantDesc is included here https://github.com/eclipse/openj9/pull/4030

DanHeidinga commented on test/functional/Java12andUp/src/org/openj9/test/java_lang/Test_String.java:
```diff
@@ -0,0 +1,68 @@
+	public void testStringDescribeConstable() {
+		String test = "test";
+
+		Optional<String> optionalTest = test.describeConstable();
```
use `.orElseThrow()`.  It simplifies the code and lets the test framework handle the unexpected exception

DanHeidinga commented on test/functional/Java12andUp/src/org/openj9/test/java_lang/Test_String.java:
```diff
@@ -0,0 +1,68 @@
+
+		if (optionalTest.isPresent()) {
+			String describedTestString = optionalTest.get();
+			Assert.assertTrue(test.equals(describedTestString));
```
should this test multiple strings?  Are there any corner cases that should be validated?  empty string?

DanHeidinga commented on test/functional/Java12andUp/src/org/openj9/test/java_lang/Test_String.java:
```diff
@@ -0,0 +1,68 @@
+		MethodHandles.Lookup lookup = MethodHandles.publicLookup();
+
+		String resolvedTest = test.resolveConstantDesc(lookup);
+		Assert.assertTrue(test.equals(resolvedTest));
```
This should test with a null lookup as well

theresa-m commented:
I've updated the tests based on the review from @DanHeidinga 

I've also added the commit from https://github.com/eclipse/openj9/pull/4104 to this pr because adding an empty testng entry does not compile.

DanHeidinga commented:
Jenkins test sanity zlinux jdk11

smlambert commented:
Jenkins test sanity zlinux jdk11

smlambert commented:
previous PR test failed with machine issue (remote call to node failed), retrying to see if transient

DanHeidinga commented:
Jenkins test sanity zlinux jdk12

DanHeidinga commented:
@theresa-m Can you rebase this PR so it'll pick up the fix for the jvm interface version # issue?

theresa-m commented:
yes. done

DanHeidinga commented:
Jenkins test sanity xlinux jdk12

DanHeidinga commented:
Jenkins test sanity xlinux jdk12

theresa-m commented:
failing tests are related to https://github.com/eclipse/openj9/issues/4661

DanHeidinga commented:
Build failures are in the `com.ibm.j9.jsr292.api.MethodHandleAPI_defineClass.test_defineClass_ClassBytes_CorruptedArray` test and unrelated to this PR
